[iOS] Added Support for XCResult

abhishekbedi1432 opened the pull request:
### Motivation
Marathon for iOS produces an amazing HTML report (in the `outputDir`) but sometimes we need the final `xcresult` bundle for many reasons (say integrating with the CI, uploading the result in repository, etc).

The `xcresult` bundles  are stored at transient paths under Xcode's `derived-data` folder. 

In this PR, we have added an enhancement of providing `xcResultBundlePath` parameter in the _Marathon_ file (under iOS Vendor Config)

As a positive side effect, this helps to avoid flooding Xcode's derived data directory.

Malinskiy commented on README.md:
```diff
@@ -6,13 +6,49 @@
 
 Please check the official [documentation](https://marathonlabs.github.io/marathon/) for installation, configuration and more
 
+## [iOS Only] Added Support to consume tags for test functions under XCTestCase Subclasses
```
Please move the docs info to the https://github.com/MarathonLabs/marathon/blob/develop/docs/_posts/2018-11-19-ios.md. They're automatically published at https://marathonlabs.github.io/marathon/

Malinskiy commented on configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt:
```diff
@@ -55,11 +55,23 @@ class ConfigurationFactory(
-                        ?: fileListProvider
+                    val resolvedResultBundlePath = marathonfileDir.resolve(configuration.vendorConfiguration.xcResultBundlePath)
+
+                    // Adding support for Test Plan
```
Please remove the comment. It doesn't make sense after we merge this.

Malinskiy commented on configuration/src/main/kotlin/com/malinskiy/marathon/config/serialization/ConfigurationFactory.kt:
```diff
@@ -55,11 +55,23 @@ class ConfigurationFactory(
                             .fileList(resolvedDerivedDataDir)
-                            .firstOrNull { it.extension == "xctestrun" }
-                        ?: throw ConfigurationException("Unable to find an xctestrun file in derived data folder")
+                            .firstOrNull { it.extension == "xctestrun" && it.name.contains("$testPlanName") } ?: throw ConfigurationException("Unable to find matching TestPlan. Please recheck if testplan is enabled")
```
You can skip creating an additional string here:
`it.name.contains("$testPlanName")` -> `it.name.contains(testPlanName)`
It might also be a good idea to use startsWith or equals instead of contains. I'm not sure if there is a contract on the file name here

Malinskiy commented on core/src/main/kotlin/com/malinskiy/marathon/Marathon.kt:
```diff
@@ -79,6 +80,10 @@ class Marathon(
         val tests = applyTestFilters(parsedTests)
         val shard = prepareTestShard(tests, analytics)
 
+        log.info("\n\n\n **** Marathon File Params **** \n")
```
This is a no-no: configuration might contain sensitive credentials, and we can't print them to the stdout

Malinskiy replied:
If you really want to do this then we need to implement some kind of masking for sensitive fields

Malinskiy commented on core/src/main/kotlin/com/malinskiy/marathon/test/Test.kt:
```diff
@@ -6,7 +6,8 @@ data class Test(
     val method: String,
-    val metaProperties: Collection<MetaProperty>
+    val metaProperties: Collection<MetaProperty>,
+    val tags: Collection<String> = emptyList()
```
MetaProprety is an abstraction over tags, annotations and so on. We should just reuse MetaProperty here

abhishekbedi1432 replied:
Refer: 
[iOS]Added Support for Cucumberish Tags / Function Annotations https://github.com/MarathonLabs/marathon/pull/702

Malinskiy commented on core/src/main/kotlin/com/malinskiy/marathon/Marathon.kt:
```diff
@@ -146,7 +151,9 @@ class Marathon(
-        return tests
+        val iosConfig = configuration.vendorConfiguration as? VendorConfiguration.IOSConfiguration
+
+        return tests.filter { it.tags.contains(iosConfig?.xcTestRunnerTag) }
```
This should just be an annotation filter, no vendor-specific logic in the core

Malinskiy commented on sample/ios-app/sample-appUITests/MoreTests.swift:
```diff
@@ -27,6 +27,7 @@ class MoreTests: XCTestCase {
         super.tearDown()
     }
 
+    // @Flowers
```
Instead of modifying existing tests we should just add more tests with these tags

Malinskiy commented on vendor/vendor-ios/src/main/kotlin/com/malinskiy/marathon/ios/IOSDevice.kt:
```diff
@@ -382,3 +390,11 @@ private fun String.toInetAddressOrNull(): InetAddress? {
 private fun TestBatch.toXcodebuildArguments(): String =
     tests.joinToString(separator = " ") { "-only-testing:\"${it.pkg}/${it.clazz}/${it.method}\"" }
+
+private fun getTimezone(): String {
```
This method sounds like getTimestamp rather than getTimezone

Malinskiy commented on vendor/vendor-ios/src/main/kotlin/com/malinskiy/marathon/ios/IOSTestParser.kt:
```diff
@@ -20,6 +21,11 @@ class IOSTestParser(private val vendorConfiguration: VendorConfiguration.IOSConf
     override suspend fun extract(): List<Test> {
+
+        if (vendorConfiguration.xcTestRunnerTag.isNullOrEmpty()) {
+            logger.warn { "[iOS] Did not find `xcTestRunnerTag` in iOS Vendor Config. This will result in running all tests" }
```
The tag filtering doesn't have anything to do with the responsibility of the parser, this probably doesn't belong here

Malinskiy commented on vendor/vendor-ios/src/main/kotlin/com/malinskiy/marathon/ios/xctestrun/Xctestrun.kt:
```diff
@@ -24,41 +24,53 @@ class Xctestrun(inputStream: InputStream) {
+*/
+
+    private val targetMap: Map<String, *> = (
+        ((((((propertyList.valueForKeypath("TestConfigurations") as? Array<Any>)?.first()) as Map<*, *>)
```
first() will throw NullPointerException here when `TestConfigurations` is missing. Reproduced on the old sample

Malinskiy reviewed (changes requested):
Logic wise we're missing pulling back the xcresult back to the marathon's side

Malinskiy commented:
Closes #620

abhishekbedi1432 commented:
Pls refer: https://github.com/MarathonLabs/marathon/pull/700 for the feature (Adding support for XCResult). I will close this PR addressing all the comments in https://github.com/MarathonLabs/marathon/pull/700
